specification/dartLangSpec.tex

```diff
@@ -20710,13 +20706,13 @@ \subsection{Static Types}
   \alt <typeNotFunction>
 
 <typeNotVoid> ::= <functionType> `?'?
-  \alt <typeNotVoidNotFunction>
+  \alt <typeNotVoidNotFunction> `?'?
 
 <typeNotFunction> ::= \VOID{}
-  \alt <typeNotVoidNotFunction>
+  \alt <typeNotVoidNotFunction> `?'?
 
-<typeNotVoidNotFunction> ::= <typeName> <typeArguments>? `?'?
-  \alt \FUNCTION{} `?'?
+<typeNotVoidNotFunction> ::= <typeName> <typeArguments>?
```

Comment:
What do we use this production for?
It's out the types allowed in extends, implements, with and mixin on types?

We could also just allow any type there, and make it an error if it's not one of the bespoke declaration kinds that you can extend. We allow you to try to extend those grammatically anyway, because of type aliases, so it might give better front-end errors if we don't make it a parse error to write `extends void`. 
It also feels arbitrary that we don't prevent `Never`.

And if we ever decide to allow extension types to implement function types, we won't need to change the grammar.

Reply:
It is used for all superinterfaces (`implements`, `extends`, `on`, and on the left hand side of `with` in mixin applications).

We could indeed allow any type, and make it a non-syntax compile-time error to have the kinds of types that we're currently ruling out in the grammar.

Most proposals have been in the opposite direction: "Adjust the grammar such that this-or-that wrong thing becomes a syntax error". Arguably, the grammar is easier to read if it describes the language properties more directly, and harder to read if it has more rules and more detail, so there is no obvious direction which is always the correct/useful one.

It is worth noting that various constructs (including type aliases and extension types) can be used to introduce types that are not allowed as superinterfaces, and yet they're syntactically identical to other types that are allowed.

However, it isn't obvious to me that it's helpful to indicate in the grammar that a function type can be a superinterface, and yet every single function type used as a superinterface is a compile-time error.

So I'll keep the grammar rule for now. Please create an issue if you wish to dive deeper into this topic.